File: Data Structures & Algorithms/string-encode-and-decode/submission-4.java
class Solution {

    public String encode(List<String> strs) {
        if(strs.isEmpty()) return "";
        StringBuilder res = new StringBuilder();
        List<Integer> sizes = new ArrayList<>();
        for(String str : strs){
            sizes.add(str.length());
        }
        
        for(int s : sizes){
            res.append(s).append(',');
        }

        res.append('#');
        for(String str :strs){
            res.append(str);
        }

        return res.toString();
    }

    public List<String> decode(String str) {
        if(str.length() == 0) return new ArrayList<>();

        List<String> res = new ArrayList<>();
        List<Integer> sizes = new ArrayList<>();
        int i = 0;
        while(str.charAt(i) != '#'){
            StringBuilder curr = new StringBuilder();
            while(str.charAt(i) != ','){
                curr.append(str.charAt(i));
                i++;
            }

            sizes.add(Integer.parseInt(curr.toString()));
            i++;
        }
        i++;
        for(int s : sizes){
            res.add(str.substring(i, i + s));
            i += s;
        }

        return res;
    }
}

File: Data Structures & Algorithms/palindrome-partitioning/submission-4.js
class Solution {
    /**
     * @param {string} s
     * @return {string[][]}
     */
    
    // aab
    // [a, a, b]
    // [aa, b]
    // [a, ab]
    // [aab]

    // [a] 
    // [a, a] [aa]
    // [a, a, b] [a, ab] [aa, b] [aab]



    partition(s) {
        const res = [];

        const substring = (curr, i) => {
            if(s.length == i) {
                for(let i = 0; i < curr.length; i++){
                    if(!this.isPalindrome(curr[i])) return;
                }
                res.push([...curr]);
                return;
            }

            curr.push(s[i]);
            substring(curr, i + 1);
            curr.pop();
            curr[curr.length - 1] = curr[curr.length - 1] + s[i];
            substring(curr, i + 1);

        }

        if(s.length == 0) return []
        substring([s[0]], 1);
        return res;
    }


    isPalindrome(s){
        let [l, r] = [0, s.length - 1];
        while(l < r) {
            if(s[l] != s[r]) return false;
            l++;
            r--;
        }
        return true;
    }
}
